### PR TITLE
fix: resolve lock length > chunks size issue

### PR DESCRIPTION
**Pull request by quake**

### CODE REVIEW

File: ckb-lock-common/src/generate_sighash_all.rs [hunk 1]
```diff
@@ -1,118 +1,79 @@
...
-    loop {
-        let sysret = load_and_hash_witness(&mut blake2b_ctx, 0, i, Source::Input, true);
-        match sysret {
-            Err(SysError::IndexOutOfBound) => break,
-            Err(x) => return Err(x.into()),
-            Ok(_) => i += 1,
-        }
-    }
+    let i = calculate_inputs_len()?;
+    load_and_hash_witness(&mut ctx, i, Source::Input);
```
Comment (XuJiandong) [inline ckb-lock-common/src/generate_sighash_all.rs:56]: Need to hash all remaining witnesses started from index `i`, not only one.

File: ckb-lock-common/src/generate_sighash_all.rs [hunk 2]
```diff
@@ -1,118 +1,79 @@
...
+                }
+            }
+            ctx.update(&chunk_data);
         }
+    } else {
+        return Err(Error::Encoding);
     }
 
+    // Digest other witnesses in the script group.
+    load_and_hash_witness(&mut ctx, 1, Source::GroupInput);
```
Comment (XuJiandong) [inline ckb-lock-common/src/generate_sighash_all.rs:53]: hash script group from index 1 to N.

File: ckb-lock-common/src/generate_sighash_all.rs [hunk 3]
```diff
@@ -129,18 +90,58 @@ fn calculate_inputs_len() -> Result<usize, Error> {
...
+                Some((loaded_len, buf))
+            }
+            Err(SysError::LengthNotEnough(actual_len)) => {
+                self.offset += self.chunk_size;
+                if self.len == 0 {
+                    self.len = actual_len;
+                }
+                Some((self.len, buf))
+            }
+            Err(SysError::IndexOutOfBound) => {
```
Comment (XuJiandong) [inline ckb-lock-common/src/generate_sighash_all.rs:140]: It's better to handle `IndexOutOfBound` outside of iterator.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: ckb-lock-common/src/generate_sighash_all.rs (hunk 1 above)
Comment (quake) [reply ckb-lock-common/src/generate_sighash_all.rs:56]: there is a loop in fn `load_and_hash_witness` to hash all witnesses starting from the index i.

**Review by XuJiandong — APPROVED** [review]

**Comment by XuJiandong** [comment]
Please merge it.

**Pull request merged**